phy/plot/_vispy_utils.py

```diff
@@ -12,175 +12,418 @@
+        return self._spike_ids
+
+    @spike_ids.setter
+    def spike_ids(self, value):
+        value = _as_array(value)
+        self._set_or_assert_n_spikes(value)
+        self._spike_ids = value
+        self.set_to_bake('spikes')
+
+    # TODO: channel_ids
+
+    @property
+    def cluster_ids(self):
+        """Clusters of the displayed spikes."""
+        return _unique(self.spike_clusters[self.spike_ids])
+
+    @property
+    def n_clusters(self):
+        return len(self.cluster_ids)
+
+    @property
+    def cluster_colors(self):
+        """Colors of the displayed clusters."""
+        return self._cluster_colors
+
+    @cluster_colors.setter
+    def cluster_colors(self, value):
+        self._cluster_colors = _as_array(value)
+        assert len(self._cluster_colors) == self.n_clusters
+        self.set_to_bake('color')
+
+    # Data baking
+    # -------------------------------------------------------------------------
+
+    def _bake_color(self):
+        u_cluster_color = self.cluster_colors.reshape((1, self.n_clusters, -1))
+        u_cluster_color = (u_cluster_color * 255).astype(np.uint8)
+        # TODO: more efficient to update the data from an existing texture
+        self.program['u_cluster_color'] = gloo.Texture2D(u_cluster_color)
+        debug("bake color", u_cluster_color.shape)
+
+    def _bake(self):
+        """Prepare and upload the data on the GPU.
+
+        Return whether something has been baked or not.
+
+        """
+        if self.n_spikes is None or self.n_spikes == 0:
+            return
+        n_bake = len(self._to_bake)
+        # Bake what needs to be baked.
+        # WARNING: the bake functions are called in alphabetical order.
+        # Tweak the names if there are dependencies between the functions.
+        for bake in sorted(self._to_bake):
+            # Name of the private baking method.
+            name = '_bake_{0:s}'.format(bake)
+            if hasattr(self, name):
+                getattr(self, name)()
+        self._to_bake = []
+        return n_bake > 0
+
+    def draw(self):
+        """Draw the waveforms."""
+        # Bake what needs to be baked at this point.
+        self._bake()
+        if self.n_spikes is not None and self.n_spikes > 0:
+            self.program.draw(self._gl_draw_mode)
+
+
+#------------------------------------------------------------------------------
+# Base spike canvas
+#------------------------------------------------------------------------------
+
+class BaseSpikeCanvas(app.Canvas):
+    _visual_class = None
+
+    def __init__(self, **kwargs):
+        super(BaseSpikeCanvas, self).__init__(keys='interactive', **kwargs)
+        self.visual = self._visual_class()
+        self._pz = PanZoom()
+        self._pz.add(self.visual.program)
+        self._pz.attach(self)
 
     def on_draw(self, event):
-        gloo.clear()
-        for visual in self.visuals:
-            visual.draw(self._tr)
+        gloo.clear('black')
+        self.visual.draw()
```

Comment:
Usually you pass a TransformSystem that has information about the different coordinate frames. If you're using the SceneCanvas you get it automatically, if not then you construct it yourself.

Check out the `TextVisual` example, which doesn't use SceneCanvas at all, to see how it can be done:

https://github.com/vispy/vispy/blob/master/examples/basics/visuals/text_visual.py

This example is designed to draw text at 48 point size (initially), i.e. 48/72 of an inch, on every screen. Then the text visual is set up to use the TransformSystem to get the right number of physical pixels to draw to get it to be that physical size:

https://github.com/vispy/vispy/blob/master/vispy/visuals/text/text.py#L464

Note that the `LineVisual` also uses the transform system to be able to scale up the `glLineWidth` call if necessary:

https://github.com/vispy/vispy/blob/master/vispy/visuals/line/line.py#L354